move-mutator/src/lib.rs

```diff
@@ -76,107 +77,138 @@ pub fn run_move_mutator(
     info!("Generated AST");
 
+    // For the next compilation steps, we don't need to fetch git deps again.
+    let mut config = config.clone();
+    config.skip_fetch_latest_git_deps = true;
+
     if mutator_configuration.project.apply_coverage {
         // This implies additional compilation inside.
         mutator_configuration
             .coverage
-            .compute_coverage(config, &package_path)?;
+            .compute_coverage(&config, &package_path)?;
     }
 
     let mutants = mutate::mutate(&env, &mutator_configuration)?;
     let output_dir = output::setup_output_dir(&mutator_configuration)?;
-    let mut report: Report = Report::new();
-
-    for mutant in &mutants {
-        let file_id = &mutant.get_file_id();
-        let source = env.get_file_source(*file_id);
-        let filename = env.get_file(*file_id);
-        let path = Path::new(filename);
-
-        trace!("Processing file: {path:?}");
-
-        let mut mutated_sources = mutant.apply(source);
 
-        // If the downsample ratio is set, we need to downsample the mutants.
+    // Generate mutants and extract all info needed for rayon threads below.
+    let mut transformed_mutants: Vec<_> = mutants
+        .into_iter()
+        .flat_map(|mutant| {
+            let file_id = mutant.get_file_id();
+            let original_source = env.get_file_source(file_id);
+            let filename = env.get_file(file_id);
+            let path = Path::new(filename)
+                .canonicalize()
+                .expect("canonicalizing failed");
+            let fn_name = mutant.get_function_name().unwrap_or_default();
+            let mod_name = mutant.get_module_name().unwrap_or("script".to_owned());
+
+            mutant
+                .apply(original_source)
+                .into_iter()
+                .map(|mutant_info| {
+                    (
+                        mutant_info,
+                        fn_name.clone(),
+                        mod_name.clone(),
+                        path.clone(),
+                        original_source,
+                    )
+                })
+                .collect::<Vec<_>>()
+        })
+        .collect();
+
+    // If the downsample ratio is set, we need to downsample the mutants.
+    if let Some(percentage) = mutator_configuration.project.downsampling_ratio_percentage {
         //TODO: currently we are downsampling the mutants after they are generated. This is not
         // ideal as we are generating all mutants and then removing some of them.
-        if let Some(percentage) = mutator_configuration.project.downsampling_ratio_percentage {
-            let no_of_mutants_to_keep = mutated_sources
-                .len()
-                .saturating_sub((mutated_sources.len() * percentage).div_ceil(100));
-            assert!(
-                no_of_mutants_to_keep <= mutated_sources.len(),
-                "Invalid downsampling ratio"
-            );
+        let total_mutants = transformed_mutants.len();
```

Comment:
Are we planning to fix this TODO about downsampling before mutant generation rather than after?

Reply:
No, the mutant generation is so fast these days that we can keep the current solution. I'll remove the TODO comment